ultraplot/ticker.py

```diff
@@ -822,6 +846,492 @@ def __call__(self, x, pos=None):  # noqa: U100
         return string
 
 
+class CFDatetimeFormatter(mticker.Formatter):
+    """
+    Format dates using `cftime.datetime.strftime` format strings.
+    """
+
+    def __init__(self, fmt, calendar="standard", units="days since 2000-01-01"):
+        """
+        Parameters
+        ----------
+        fmt : str
+            The `strftime` format string.
+        calendar : str, default: 'standard'
+            The calendar for interpreting numeric tick values.
+        units : str, default: 'days since 2000-01-01'
+            The time units for interpreting numeric tick values.
+        """
+        if cftime is None:
+            raise ModuleNotFoundError("cftime is required for CFDatetimeFormatter.")
+        self._format = fmt
+        self._calendar = calendar
+        self._units = units
+
+    def __call__(self, x, pos=None):
+        if isinstance(x, cftime.datetime):
+            dt = x
+        else:
+            dt = cftime.num2date(x, self._units, calendar=self._calendar)
+        return dt.strftime(self._format)
+
+
+class AutoCFDatetimeFormatter(mticker.Formatter):
+    """Automatic formatter for `cftime.datetime` data."""
+
+    def __init__(self, locator, calendar, time_units=None):
+        self.locator = locator
+        self.calendar = calendar
+        self.time_units = time_units or _TIME_UNITS
+
+    def pick_format(self, resolution):
+        return _RESOLUTION_TO_FORMAT[resolution]
+
+    def __call__(self, x, pos=0):
+        format_string = self.pick_format(self.locator.resolution)
+        if isinstance(x, cftime.datetime):
+            dt = x
+        else:
+            dt = cftime.num2date(x, self.time_units, calendar=self.calendar)
+        return dt.strftime(format_string)
+
+
+class AutoCFDatetimeLocator(mticker.Locator):
+    """Determines tick locations when plotting `cftime.datetime` data."""
+
+    real_world_calendars = (
+        "gregorian",
+        "julian",
+        "proleptic_gregorian",
+        "standard",
+    )
```

Comment:
Can we get these from cftime automatically somehow? Otherwise we have to update them if they go out of sync and that will cause a bug eventually.

Reply:
There is no public api for this. We could use `_calendars`but thay may be brittle. I now look for the calendar attribute.